### ISSUE TITLE
How to implement Windows Auth on BlazorClient ?

### ISSUE THREAD

**Issue by drakeforte5** [MarimerLLC/cslaforum#825]
I'm tracing the csla source code and I get lost as I came from the version 2.xx which was more straightforward to debug/follow, eventually upgraded. Recent versions are much more powerful and complex hence I need to ask the question.

HttpProxy, which is what the BlazorClient uses seem to retain the old-windows convention of not serializing the principal if AuthenticationType == "Windows".  However the server-counterpart which I believe is the HttpPortalController/HttpPortal seems to just deserialize the principal.  

I added Authorize attribute on the BlazorExample.Server.Controller.DataPortalController to test it out while switching the Enable Windows Auth on the server project.

`[Route("api/[controller]")]

  [ApiController]

  [Authorize()]

  public class DataPortalController : Csla.Server.Hosts.HttpPortalController`

Also, I added this to the BlazorExample.Client to force windows auth.

`public void Configure(IComponentsApplicationBuilder app)
    {
      app.AddComponent<App>("app");

      app.UseCsla((c) => 
        c.DataPortal().DefaultProxy(typeof(Csla.DataPortalClient.HttpProxy), "/api/DataPortal").
        **AuthenticationType("Windows")**
        );
    }`

I am getting an error 500 and the Csla.ApplicationContext.User is null when debugged on the server-side.

Is there a setting or code that I need to implement? I was hoping that the HttpClient instance on the HttpProxy would just use  HttpHandler with .UseDefaultCredentials when using Windows auth and that the client request would just work seamlessly.  (https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.usedefaultcredentials?view=netframework-4.8) 

Angular's HttpRequest has a similar concept where if you set withCredentials = true then the windows identity will get passed into to the client's request to the server. (https://angular.io/api/common/http/HttpRequest)

Please advise.

Thank you.

**Comment by drakeforte5** [comment]
I verified the BlazorExample.Server nuget packages and it does reference Csla.AspNetCore and I am setting the AuthenticationType to Windows:

`      app.UseCsla((c) =>
        c.DataPortal().AuthenticationType("Windows"));`

but still the Csla.ApplicationContext.User is null hence I added this line in the DataPortalController.PostAsync to solve the issue.
`      Csla.ApplicationContext.User = HttpContext.User;`

The code is from the CSLA Blazor sample just modified to use Windows Auth.  Am I missing something else, some other setting perhaps?

**Comment by drakeforte5** [comment]
One more observation.  While inside the DataportalController.PosAsync I typed this in the VS debug command window:

?  System.Type.GetType("Csla.Web.ApplicationContextManager, Csla.Web")

and I get null which translates to a null value for the Csla.ApplicationContext.WebContextManager .

This is supposedly mimicking line number 55 in the Csla.Shared.ApplicationContext where it is resolving the WebContextManager.

**Comment by drakeforte5** [comment]
This might be bug Rocky.  

This returns null: ? System.Type.GetType("Csla.Web.ApplicationContextManager, **Csla.Web**")

While this returns the correct type, it is because of the assembly name:
? System.Type.GetType("Csla.Web.ApplicationContextManager, **Csla.AspNetCore.Mvc**")

**Comment by rockfordlhotka** [comment]
There may be a bug. But first, I think you are misunderstanding the way `AuthenticationType` works. This is described in the _Using CSLA 4_ data portal book.

That confg value must be the same on client and server, because it controls how (or if) the data portal transports the user's identity from client to server on each data portal call.

* `Csla` means that the data portal transports the client-side principal to the server with each call, and sets the server's principal to match that from the client
* `Windows` means that the data portal does _not_ transport the client-side principal to the server, and does _not_ set the server-side principal at all

The idea behind setting the value to `Windows` is that the user's identity is managed by something outside CSLA - originally Windows, but these days there are probably other options. The point being, if the value is set to `Windows` then the server (Windows or _your code_) are responsible for setting the principal before the data portal is invoked.

**Comment by drakeforte5** [comment]
Thanks Rocky. Yes the AuthenticationType upon reviewing the source code has
retained the concept when set to Windows ever since Csla 2.x and it looks
like there might be a bug on the ApplicationContext just like I described.
For now I am setting the ApplicationContext.User to Httpcontext.User to
remediate the issue on the DaportalController.

**Comment by rockfordlhotka** [comment]
@drakeforte5 totally agree with the bug. On it! 😄

**Comment by rockfordlhotka** [comment]
I just pushed a prerelease of v5.1.0 that should fix the issue.

**Comment by drakeforte5** [comment]
I'll go ahead and test it and close this ticket after. Thanks.

**Comment by drakeforte5** [comment]
It is now working and the user is now being returned by Csla.ApplicationContext.User on pre-v5.1.0.  A very important point however is that the HttpContextAccessor has to be injected during the  BlazorExample.Server.Startup.ConfigureServices().

` services.AddHttpContextAccessor();`

I have conflicting information about the HttpContextAccessor as I thought it is now automatically injected but apparently not because If I don't include the AddHttpContextAccessor the Csla.Application.User returns UnauthenticatedPrincipal naturally because Csla.AspNetCore.Shared.ApplicationContextManager.HttpContext line 51:

`  var httpContextAccessor = (IHttpContextAccessor)_serviceProvider.GetService(typeof(IHttpContextAccessor));`  

will be null. 

Anyway, it is now working with a caveat that AddHttpContextAccessor() must be called on the Startup.

Thanks!

**Comment by rockfordlhotka** [comment]
Yes, I found similarly confusing info. But what I gather is the latest/best approach is that if you want to access `HttpContext` (or if you want CSLA to be able to access it) in any aspnetcore server project, you need the following in `Startup.ConfigureServices`:

```c#
      services.AddHttpContextAccessor();
      services.AddCsla();
```

And then in `Startup.Configure`:

```c#
      app.UseCsla();
      Configuration.ConfigureCsla();
```

Inside of CSLA the `IHttpContextAccessor` is used as you describe, but it relies on the `AddHttpContextAccessor` method being called in `ConfigureServices`.
